compiler/rustc_index/src/vec.rs

```diff
@@ -138,12 +150,20 @@ macro_rules! newtype_index {
                 }
             }
 
+            /// Creates a new index from a given `u32`.
+            ///
+            /// # Safety
+            ///
+            /// This function is unsafe because initializing a layout restricted type's
+            /// field with a value outside the valid range is undefined behavior.
+            ///
+            /// Prefer using `from_u32`.
             #[inline]
             $v const unsafe fn from_u32_unchecked(value: u32) -> Self {
                 Self { private: value }
```

Comment:
This can't be true, right? Just constructing the index can't cause UB, only using it in an invalid way afterwards. Could you look in this module for where it depends on the invariant that `value < MAX`?

Reply:
It is because of `rustc_layout_scalar_valid_range_end`.

I think better docs here would be:
```
# Safety

The provided value must be less than or equal to the max value for the newtype. Providing a value outside this range is undefined due to layout restrictions.
```